File: rust/everybody-codes/src/bin/2025/y25d06.rs
use everybody_codes::utilities::inputs::get_event_inputs;
use utilities::structs::stopwatch::{ReportDuration, Stopwatch};

type Input<'a> = &'a str;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let (input1, input2, input3) = get_event_inputs(25, 6);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input1), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input2), stopwatch.lap().report());
    println!("3. {} ({})", part3(&input3), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn part1(input: Input) -> usize {
    let mut knights = 0;
    let mut pairs = 0;
    for &b in input.as_bytes() {
        match b {
            b'A' => knights += 1,
            b'a' => pairs += knights,
            _ => {}
        }
    }
    pairs
}

fn part2(input: Input) -> usize {
    let mut knights: [usize; 3] = [0; 3];
    let mut pairs = 0;
    for &b in input.as_bytes() {
        match b {
            b if b.is_ascii_uppercase() => { knights[(b - b'A') as usize] += 1; },
            b => { pairs += knights[(b - b'a') as usize]; },
        }
    }
    pairs
}

fn part3(input: Input) -> usize {
    let mut pre_repeat: [usize; 3] = [0; 3];
    let mut pre: [usize; 3] = [0; 3];
    let mut post_repeat: [usize; 3] = [0; 3];
    let mut post: [usize; 3] = [0; 3];
    let mut pairs = 0;

    // constants
    let input = input.as_bytes();
    let distance = 1000;
    let repeats = 1000;

    //pre-fill
    for &b in &input[input.len() - distance..] {
        if b.is_ascii_uppercase() {
            pre_repeat[(b - b'A') as usize] += 1;
        }
    }
    for &b in &input[0..=distance] {
        if b.is_ascii_uppercase() {
            post[(b - b'A') as usize] += 1;
        }
    }

    for (idx, &b) in input.iter().enumerate() {
        if b.is_ascii_uppercase() {
            let cur_idx = (b - b'A') as usize;
            post[cur_idx] -= 1;
        } else {
            let b_idx = (b - b'a') as usize;
            pairs += pre_repeat[b_idx] * (repeats - 1)
                + pre[b_idx] * repeats
                + post[b_idx] * repeats
                + post_repeat[b_idx] * (repeats - 1);
        }
        // drop stuff behind
        let pre_idx = idx as isize - distance as isize ;
        if pre_idx < 0 {
            let pre_drop = input[pre_idx.rem_euclid(input.len() as isize) as usize];
            if pre_drop.is_ascii_uppercase() {
                pre_repeat[(pre_drop - b'A') as usize] -= 1;
            }
        } else {
            let pre_drop = input[pre_idx as usize];
            if pre_drop.is_ascii_uppercase() {
                pre[(pre_drop - b'A') as usize] -= 1;
            }
        }
        let post_idx = idx + distance + 1;
        if post_idx >= input.len() {
            let post_drop = input[post_idx.rem_euclid(input.len())];
            if post_drop.is_ascii_uppercase() {
                post_repeat[(post_drop - b'A') as usize] += 1;
            }
        } else {
            let post_drop = input[post_idx];
            if post_drop.is_ascii_uppercase() {
                post[(post_drop - b'A') as usize] += 1;
            }
        }
        if b.is_ascii_uppercase() {
            let cur_idx = (b - b'A') as usize;
            pre[cur_idx] += 1;
        }
    }
    pairs
}

#[test]
fn default() {
    let (input1, input2, input3) = get_event_inputs(25, 6);
    assert_eq!(190, part1(&input1));
    assert_eq!(4011, part2(&input2));
    assert_eq!(1665939853, part3(&input3));
}

// Input parsed (34μs)
// 1. 190 (7μs)
// 2. 4011 (2μs)
// 3. 1665939853 (118μs)
// Total: 165μs

File: kotlin/everybody_codes/src/main/kotlin/org/gristle/pdxpuzzles/everybodycodes/y2025/Y25D06.kt
package org.gristle.pdxpuzzles.everybodycodes.y2025

import org.gristle.pdxpuzzles.everybodycodes.utilities.Day

object Y25D06 : Day {
    override fun part1(input: String): Int {
        var knights = 0
        var pairs = 0
        for (c in input) {
            when (c) {
                'A' -> knights++
                'a' -> pairs += knights
                else -> {}
            }
        }
        return pairs
    }
    override fun part2(input: String): Int {
        val knights = IntArray(3)
        var pairs = 0
        for (c in input) {
            when {
                c.isUpperCase() -> knights[c - 'A'] += 1
                else -> pairs += knights[c - 'a']
            }
        }
        return pairs
    }
    override fun part3(input: String): Int {
        val preRepeat = IntArray(3)
        val pre = IntArray(3)
        val postRepeat = IntArray(3)
        val post = IntArray(3)
        var pairs = 0

        // constants
        val distance = 1000
        val repeats = 1000

        // pre-fill
        for (c in input.takeLast(distance)) {
            if (c.isUpperCase()) {
                preRepeat[c - 'A']++
            }
        }
        for (c in input.take(distance)) {
            if (c.isUpperCase()) {
                post[c - 'A']++
            }
        }

        // main loop
        for ((idx, c) in input.withIndex()) {
            if (c.isUpperCase()) {
                post[c - 'A']--
            } else {
                val cIdx = c - 'a'
                pairs += preRepeat[cIdx] * (repeats - 1) +
                        pre[cIdx] * repeats +
                        post[cIdx] * repeats +
                        postRepeat[cIdx] * (repeats - 1)
            }

            //drop stuff behind
            val preIdx = idx - distance
            if (preIdx < 0) {
                val preDrop = input[preIdx.mod(input.length)]
                if (preDrop.isUpperCase()) {
                    preRepeat[preDrop - 'A']--
                }
            } else {
                val preDrop = input[preIdx]
                if (preDrop.isUpperCase()) {
                    pre[preDrop - 'A']--
                }
            }
            val postIdx = idx + distance + 1
            if (postIdx >= input.length) {
                val postDrop = input[postIdx % input.length]
                if (postDrop.isUpperCase()) {
                    preRepeat[postDrop - 'A']++
                }
            } else {
                val postDrop = input[postIdx]
                if (postDrop.isUpperCase()) {
                    pre[postDrop - 'A']++
                }
            }
            if (c.isUpperCase()) {
                pre[c - 'A']++
            }
        }
        return pairs
    }
}

fun main() = Day.runDay(Y25D06::class)

//    Quest 1: 190 (0ms)
//    Quest 2: 4011 (3ms)
//    Quest 3: 1665939853 (3ms)
//    Total time: 6ms
